knowledgebase/profiling-clickhouse-with-llvm-xray.md

```diff
@@ -0,0 +1,139 @@
+---
+date: 2024-11-13
+---
+
+# Profiling ClickHouse with LLVM's XRay
+
+## Types of profilers
+
+LLVM already includes a tool that instruments the code that allows us to do [instrumentation
+profiling](https://en.wikipedia.org/wiki/Profiling_(computer_programming)#Instrumentation). As
+opposed to [sampling or statistical
+profiling](https://en.wikipedia.org/wiki/Profiling_(computer_programming)#Statistical_profilers),
+it's very precise without losing any calls, at the expense of needing to instrument the code and be
+more resource expensive.
+
+In a few words, an instrumentation profiler introduces new code to track the call to all functions.
+Statistical profilers allow us to run the code without requiring any changes, taking snapshots
+periodically to see the state of the application. So, only the functions running while the snapshot
+is taken are considered. [perf](https://en.wikipedia.org/wiki/Perf_%28Linux%29) is a very well-known
+statistical profiler.
+
+## How to profile with XRay
+
+### Instrument the code
+
+Image the following souce code:
+
+```c++
```

Comment:
Syntax hightlighing in https://clickhouse.com/docs/knowledgebase/profiling-clickhouse-with-llvm-xray is missing for this code block. I think you'll need to use `cpp` here.

Reply:
Yep, I noticed that but I thought our markdown thingy on the web didn't support C++ syntax :sweat_smile: 
Will do, thanks for the tip